File: string/1456_Maximum_Number_of_Vowels_in_a_Substring_of_Given_Length/1456_Maximum_Number_of_Vowels_in_a_Substring_of_Given_Length.go
package main

import "fmt"

func maxVowels(s string, k int) int {
	res := 0
	dp := make([]int, len(s)+1)
	count := 0
	for i := 0; i < len(s); i++ {
		dp[i] = count
		if check(s[i]) == true {
			count++
		}
	}
	dp[len(s)] = count
	for i := k; i <= len(s); i++ {
		if dp[i]-dp[i-k] > res {
			res = dp[i] - dp[i-k]
		}
	}
	return res
}

func check(str byte) bool {
	if str == 'a' || str == 'e' || str == 'i' || str == 'o' || str == 'u' {
		return true
	}
	return false
}

func main() {
	s := "abciiidef"
	k := 3
	res := maxVowels(s, k)
	fmt.Println(res)
}

File: string/1456_Maximum_Number_of_Vowels_in_a_Substring_of_Given_Length/1456_Maximum_Number_of_Vowels_in_a_Substring_of_Given_Length.py

class Solution:
    def maxVowels(self, s: str, k: int) -> int:
        d = {'a': 0, 'e': 0, 'i': 0, 'o': 0, 'u': 0}
        for c in s[:k]:
            if c in d:
                d[c] += 1
        res = sum(d.values())
        for i in range(k, len(s)):
            if s[i-k] in d:
                d[s[i-k]] -= 1
            if s[i] in d:
                d[s[i]] += 1
            res = max(res, sum(d.values()))
        return res


s = "abciiidef"
k = 3
res = Solution().maxVowels(s, k)
print(res)